### PR TITLE
Network Security Group API and Network Security Group Rule API

### PR DESCRIPTION
**Pull request by jmspring**
@nacx This pull request implements the REST API calls for Azure Resource Manager Network Security Group and Network Security Rule APIs (they do intersect).  This PR has no dependencies on any PRs outside what is currently in master for jclouds-labs.

### CODE REVIEW

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityRuleApi.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   @Path("/securityRules/{networksecurityrulename}")
+   @GET
+   @Fallback(NullOnNotFoundOr404.class)
+   NetworkSecurityRule get(@PathParam("networksecurityrulename") String ruleName);
+
+   @Named("networksecurityrule:delete")
+   @Path("/securityRules/{networksecurityrulename}")
+   @DELETE
+   @ResponseParser(FalseOn204.class)
+   boolean delete(@PathParam("networksecurityrulename") String ruleName);
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityRuleApi.java]: Since this delete is async, consider using `URIParser`, similar to [this](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/ResourceGroupApi.java#L89-L94)

Please also test the returned uri to ensure it is not null. Also add the fallback test case like [this](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/ResourceGroupApiMockTest.java#L141-L148)

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+   @GET
+   @SelectJson("value")
+   @Fallback(EmptyListOnNotFoundOr404.class)
+   List<NetworkSecurityGroup> list();
+
+   @Named("networksecuritygroup:delete")
+   @Path("/{networksecuritygroupname}")
+   @DELETE
+   @ResponseParser(FalseOn204.class)
+   boolean delete(@PathParam("networksecuritygroupname") String nsgName);
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java]: Since this delete is async, consider using `URIParser`, similar to [this](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/ResourceGroupApi.java#L89-L94)

Please also test the returned uri to ensure it is not null. Also add the fallback test case like [this](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/ResourceGroupApiMockTest.java#L141-L148)

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+   @Named("networksecuritygroup:create_or_update")
+   @Path("/{networksecuritygroupname}")
+   @PUT
+   @MapBinder(BindToJsonPayload.class)
+   @Fallback(NullOnNotFoundOr404.class)
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java]: Don't use 404 fallbacks in POST/PUT

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityRuleApi.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+@QueryParams(keys = "api-version", values = "2016-03-30")
+@RequestFilters(OAuthFilter.class)
+@Consumes(MediaType.APPLICATION_JSON)
+public interface NetworkSecurityRuleApi {
+   @Named("networksecurityrule:create_or_update")
+   @Path("/securityRules/{networksecurityrulename}")
+   @PUT
+   @MapBinder(BindToJsonPayload.class)
+   @Fallback(NullOnNotFoundOr404.class)
```
Comment (ritazh) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityRuleApi.java]: Don't use 404 fallbacks in POST/PUT

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java (hunk 3 above)
Comment (jmspring) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java]: Specific example for the 404s?  Thanks.

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java (hunk 2 above)
Comment (jmspring) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/NetworkSecurityGroupApi.java]: I'll take a look

**Comment by jmspring** [comment]
URI used for delete, tests updated.

**Comment by jmspring** [comment]
Redoing due to PR / tree getting into a bad state.
